File: C_Codigos/filtragens_espaciais/filtro_da_mediana.c
#include <opencv2/opencv.hpp>
#include <iostream>
#include <filesystem>
#include <chrono>
#include <vector>
#include <algorithm>

using namespace std;
using namespace cv;
namespace fs = std::filesystem;

void filtroDaMediana(const Mat& imagem, Mat& imagemFiltrada, int tamJanela = 3) {
    int altura = imagem.rows;
    int largura = imagem.cols;
    int canais = imagem.channels();
    int margem = tamJanela / 2;

    imagemFiltrada = imagem.clone();

    for (int i = margem; i < altura - margem; ++i) {
        for (int j = margem; j < largura - margem; ++j) {
            for (int canal = 0; canal < canais; ++canal) {
                vector<uchar> regiao;
                for (int x = -margem; x <= margem; ++x) {
                    for (int y = -margem; y <= margem; ++y) {
                        regiao.push_back(imagem.at<Vec3b>(i + x, j + y)[canal]);
                    }
                }
                nth_element(regiao.begin(), regiao.begin() + regiao.size() / 2, regiao.end());
                uchar mediana = regiao[regiao.size() / 2];
                imagemFiltrada.at<Vec3b>(i, j)[canal] = mediana;
            }
        }
    }
}

vector<double> processarDiretorio(const string& input_dir, const string& output_dir, int tamJanela = 3) {
    vector<double> tempos_execucao;

    for (const auto& entry : fs::directory_iterator(input_dir)) {
        if (entry.path().extension() == ".png" || entry.path().extension() == ".jpg" || 
            entry.path().extension() == ".jpeg" || entry.path().extension() == ".bmp") {

            Mat imagem = imread(entry.path().string(), IMREAD_COLOR);
            if (!imagem.empty()) {
                Mat imagemFiltrada;
                auto start_time = chrono::high_resolution_clock::now();
                filtroDaMediana(imagem, imagemFiltrada, tamJanela);
                auto end_time = chrono::high_resolution_clock::now();
                chrono::duration<double, milli> elapsed_time = end_time - start_time;
                tempos_execucao.push_back(elapsed_time.count());

                string caminho_imagem_filtrada = output_dir + "/" + entry.path().filename().string();
                imwrite(caminho_imagem_filtrada, imagemFiltrada);
                cout << "Tempo de processamento de " << entry.path().filename().string() << ": " << elapsed_time.count() << " ms" << endl;
            } else {
                cout << "Erro ao carregar a imagem: " << entry.path().string() << endl;
            }
        }
    }

    return tempos_execucao;
}

void multiplasExecucoes(const string& input_dir, const string& output_dir, int tamJanela = 3, int execucoes = 1) {
    vector<double> tempos_todas_execucoes;
    for (int execucao = 0; execucao < execucoes; ++execucao) {
        cout << "Iniciando execução " << execucao + 1 << endl;
        vector<double> tempos_execucao = processarDiretorio(input_dir, output_dir, tamJanela);
        tempos_todas_execucoes.insert(tempos_todas_execucoes.end(), tempos_execucao.begin(), tempos_execucao.end());
        if (!tempos_execucao.empty()) {
            double media_execucao = accumulate(tempos_execucao.begin(), tempos_execucao.end(), 0.0) / tempos_execucao.size();
            cout << "Média de tempo para a execução " << execucao + 1 << ": " << media_execucao << " ms" << endl;
        } else {
            cout << "Nenhuma imagem processada nesta execução." << endl;
        }
    }

    if (!tempos_todas_execucoes.empty()) {
        double media_geral = accumulate(tempos_todas_execucoes.begin(), tempos_todas_execucoes.end(), 0.0) / tempos_todas_execucoes.size();
        cout << "Média geral das médias de tempo após " << execucoes << " execuções: " << media_geral << " ms" << endl;
    } else {
        cout << "Nenhuma imagem foi processada em nenhuma execução." << endl;
    }
}

int main() {
    string input_dir = "C:/Users/Cliente/Downloads/base_dados/Imagens_Selecionadas";
    string output_dir = "C:/Users/Cliente/Downloads/base_dados/Saida_Python_Filtro_Mediana";

    // Chama a função que executa o processamento múltiplas vezes
    multiplasExecucoes(input_dir, output_dir);

    return 0;
}

File: Python_Codigos/filtragens_espaciais/filtro_da_mediana.py
import cv2
import numpy as np
import os
import time

def filtro_da_mediana(imagem, tamJanela=3):
    if len(imagem.shape) == 2:
        altura, largura = imagem.shape
        canais = 1
        imagem = imagem[:, :, np.newaxis]
    else:
        altura, largura, canais = imagem.shape

    imagemFiltrada = np.copy(imagem)
    margem = tamJanela // 2

    for i in range(margem, altura - margem):
        for j in range(margem, largura - margem):
            for canal in range(canais):
                regiao = imagem[i - margem:i + margem + 1, j - margem:j + margem + 1, canal]
                mediana = np.median(regiao)
                imagemFiltrada[i, j, canal] = mediana

    if canais == 1:
        imagemFiltrada = imagemFiltrada[:, :, 0]

    return imagemFiltrada

def processarDiretorio(input_dir, output_dir, tamJanela=3):
    tempos_execucao = []
    for filename in os.listdir(input_dir):
        if filename.lower().endswith(('.png', '.jpg', '.jpeg', '.bmp')):
            caminho_imagem = os.path.join(input_dir, filename)
            imagem = cv2.imread(caminho_imagem, cv2.IMREAD_COLOR)
            if imagem is not None:
                start_time = time.time()
                imagemFiltrada = filtro_da_mediana(imagem, tamJanela)
                elapsed_time = time.time() - start_time
                tempos_execucao.append(elapsed_time * 1000)

                caminho_imagem_filtrada = os.path.join(output_dir, filename)
                cv2.imwrite(caminho_imagem_filtrada, imagemFiltrada)
                print(f"Tempo de processamento de {filename}: {elapsed_time * 1000:.4f} ms")
            else:
                print(f"Erro ao carregar a imagem: {caminho_imagem}")
    return tempos_execucao

def multiplasExecucoes(input_dir, output_dir, tamJanela=3, execucoes=1):
    tempos_todas_execucoes = []
    for execucao in range(execucoes):
        print(f"Iniciando execução {execucao + 1}")
        tempos_execucao = processarDiretorio(input_dir, output_dir, tamJanela)
        tempos_todas_execucoes.extend(tempos_execucao)
        if tempos_execucao:
            media_execucao = sum(tempos_execucao) / len(tempos_execucao)
            print(f"Média de tempo para a execução {execucao + 1}: {media_execucao:.4f} ms")
        else:
            print("Nenhuma imagem processada nesta execução.")

    if tempos_todas_execucoes:
        media_geral = sum(tempos_todas_execucoes) / len(tempos_todas_execucoes)
        print(f"Média geral das médias de tempo após {execucoes} execuções: {media_geral:.4f} ms")
    else:
        print("Nenhuma imagem foi processada em nenhuma execução.")

input_dir = '/mnt/c/Users/Cliente/Downloads/base_dados/Imagens_Selecionadas'
output_dir = '/mnt/c/Users/Cliente/Downloads/base_dados/teste'

# Chama a função que executa o processamento múltiplas vezes
multiplasExecucoes(input_dir, output_dir, tamJanela=3, execucoes=1)
